src/thing_creature.c

```diff
@@ -1241,19 +1255,162 @@ void process_thing_spell_teleport_effects(struct Thing *thing, struct CastedSpel
+                                    battleid = 1;
+                                }
+                                count++;
+                                struct CreatureBattle* battle = creature_battle_get(i);
+                                if ( (battle->fighters_num != 0) && (battle_with_creature_of_player(thing->owner, i)) )
+                                {
+                                    struct Thing* tng = thing_get(battle->first_creatr);
+                                    TRACE_THING(tng);
+                                    if (creature_can_navigate_to(thing, &tng->mappos, NavRtF_NoOwner))
+                                    {
+                                        pos.x.val = tng->mappos.x.val;
+                                        pos.y.val = tng->mappos.y.val;
+                                        battleid = i + 1;
+                                        break;
+                                    }
+                                }
+                                if (count >= BATTLES_COUNT)
+                                {
+                                    battleid = 1;
+                                    break;
+                                }
+                                if (i >= BATTLES_COUNT)
+                                {
+                                    i = 0;
+                                    battleid = 1;
+                                    continue;
+                                }
+                            }
+                        }
+                        else
+                        {
+                            allowed = false;
+                        }
+                        break;
+                    }
+                    case 17:
+                    {
+                        room = room_get(cctrl->last_work_room_id);
+                        break;
+                    }
+                    case 18:
+                    {
+                        struct Coord3d cta_pos;
+                        cta_pos.x.val = subtile_coord_center(dungeon->cta_stl_x);
+                        cta_pos.y.val = subtile_coord_center(dungeon->cta_stl_y);
+                        cta_pos.z.val = subtile_coord(1,0);
+                        if (creature_can_navigate_to(thing, &cta_pos, NavRtF_NoOwner))
+                        {
+                            pos = cta_pos;
+                        }
+                        else
+                        {
+                            allowed = false;
+                        }
+                        break;
+                    }
+                    case 19:
+                    {
+                        desttng = find_hero_door_hero_can_navigate_to(thing);
+                        allowed = (!thing_is_invalid(desttng));
+                        break;
+                    }
+                    default:
+                    {
+                        rkind = zoom_key_room_order[TeleDest-1];
```

Comment:
I think it's wise to set some conditions here,... only do this if TeleDest-1 would return a room. Or check that there's no way this can go wrong.

Reply:
It shouldn't go wrong, as zoom_key_order is hard-coded. Do you think it should reset to 0 if the resulting value is not a room?